### ISSUE TITLE
Slight ergonomics wrinkle with `self` by-value parameters

### ISSUE THREAD

**Issue by ijackson** [davidpdrsn/extend#16]
Hi.  I have just found this crate.  Oh wow!  So convenient!  Thank you.

I have this code:

```
#[ext(pub, name=LocalFileExt, supertraits=Sized)]
impl fs::File {
  #[throws(io::Error)]
  fn close(self) {
```

The `supertraits=` is needed because otherwise `ext` generates a trait with a method which takes `self` but without a `Sized` bound.  Obviously that bound is satisfied in the actual impl.  Maybe `ext` could generate a `where Self: Sized` bound on each method taking `self` (ie, by value) as a receiver?

(The `#[throws]` is from fehler of course.  So excellent that `extend` and `fehler` work nicely together.)

**Comment by davidpdrsn** [comment]
Hm I'm not totally convinced but I think manually adding `where Self: Sized` on methods as necessary is the best approach. Adding that automatically based on the receiver seems feels a bit magical. I'm wondering if there are places where you wouldn't want that.

I guess it would also be a breaking change and require a major version bump.

**Comment by ijackson** [comment]
If you are extending a trait, you might very well want your extension methods to be useable on trait objects.  Likewise you might want to extend an unsized type.  Eg I have in my project an extension method on `str`, and several utility methods on one of my own traits which is normally seen as a trait object.  So always adding a `Sized` bound isn't right.

There is the well-known trick to making a trait object-safe by restricting the non-object-safe methods with `where Self: Sized`, thus producing the object-safe subset of the original trait.  It seems to me that when the user of `#[ext]` is extending a generic type (or trait), or unsized type, they can decide for themselves which things need a `Sized` bound (and whether that's best done on each method or as a supertrait for the extension trait).  Such a user is likely to understand the compiler errors and know what to do.

It is where `#[ext]` is being used to extend a concrete type that I think there's an issue.  Obviously we know it is making a trait behind the scenes, but the syntax (eg in my example above) looks entirely concrete and doesn't contain anything that isn't plainly `Sized`.  The problem occurs when the type `Self` (which in the trait is `?Sized`) is used in an argument or return value - and ony when that type is spelled `Self`.  Eg, an extension method on `File` which is ritten `fn wibble(&self) -> File` is fine without additional bounds whereas `fn wibble(&self) -> Self` breaks.

When a concrete type is being extended, it an be named explicitly everywhere but method receivers and indeed doing so is quite natural.  But method receivers have to be variants on `self`.  So I think that's why `self` deserves special treatment.

It would be better if it were possible to somehow determine whether the type being extended is necessarily `Sized` and add that bound.  But I think that's probably very hard if not entirely impossible.

Spotting the by-value `self` receiver solves the problem in the most obvious cases (and it would be a non-breaking change - anyone who already has such extension methods must already have `Sized` as a supertrait or method bound so adding another one is harmless).

**Comment by davidpdrsn** [comment]
I still feel that having `ext` magically add bounds the user didn't explicitly ask for is too magical. I also think the workarounds are fine and don't doesn't really warrant the additional complexity this would add.

**Comment by ijackson** [comment]
> I still feel that having `ext` magically add bounds the user didn't explicitly ask for is too magical

I can see why it feels that way.

Note though that the user would see almost no effects of these bounds.  Adding them would only cause code to compile which would otherwise be rejected; they would never break anything that would otherwise work.  Nor would they ever appear in any error messages.  So insofar as it's magical, it's invisible magic that just makes it DTRT.

The only way the user could tell this was happening is that sometimes they wouldn't need to write their own `Sized` bound, because `#[ext]` had added one already.  Extension methods that are non-object-safe would need a user-provided `Sized` bound precisely when they don't take `self` by value - rather than as now always needing a user-provided `Sized` bound.

> I also think the workarounds are fine and don't doesn't really warrant the additional complexity this would add.

Fair enough.  Thanks for your attention.

**Comment by davidpdrsn** [comment]
I think we should keep this issue open though. Might be something we want to revisit in the future.

**Comment by TCROC** [comment]
I just found this crate and I can agree with the original poster that this is quite amazing!  If we do decided to do something for ``supertraits=Sized``, what if instead of releasing a breaking change, we added a new trait called ``#ext_sized``?

**Comment by davidpdrsn** [comment]
I like `ext_sized` as an explicit opt-in!

**Comment by davidpdrsn** [comment]
I added `#[ext_sized]` in https://github.com/davidpdrsn/extend/pull/19. @TCROC @ijackson are you able to give it a shot and see if it works for you?

**Comment by TCROC** [comment]
I'll give it a shot when I get a spare moment.  I'm new to Rust still, so this will be better suited for @ijackson to test since he knows how its supposed to work.

**Comment by davidpdrsn** [comment]
I went ahead and shipped 1.1.0 which includes this. Let me know if you have any issues with it.

**Comment by TCROC** [comment]
Ur the man! 💪
